src/Microsoft.ML.FastTree/doc.xml

```diff
@@ -73,6 +73,64 @@
+        Instead of consuming the model's output, we could go deeper, and extract the 'intermediate outputs' that are used to produce the final score. </para>
+        There's a number of famous or popular examples of this technique:
+        <list>
+          <item>
+            <description>A deep neural net trained on the ImageNet dataset, with the last layer removed, is commonly used to compute the 'projection' of the image into the 'semantic feature space'.
+            It is observed that the Euclidian distance in this space often correlates with the 'semantic similarity': that is, all pictures of pizza are located close together,
+            and far away from pictures of kittens. </description>
+          </item>
+          <item>
+            <description>A matrix factorization and/or LDA model is also often used to extract the 'latent topics' or 'latent features' associated with users and items.</description>
+          </item>
+          <item>
+            <description>The weights of the linear model are often used as a crude indicator of 'feature importance'. At the very minimum, the 0-weight features are not needed by the model,
+            and there's no reason to compute them. </description>
+          </item>
+        </list>
+        <para>Tree featurizer uses the decision tree ensembles for feature engineering in the same fashion as above.</para>
+        <para>Let's assume that we've built a tree ensemble of 100 trees with 100 leaves each (it doesn't matter whether boosting was used or not in training). 
+        If we associate each leaf of each tree with a sequential integer, we can, for every incoming example x, 
+        produce an indicator vector L(x), where Li(x) = 1 if the example x 'falls' into the leaf #i, and 0 otherwise.</para>
+        <para>Thus, for every example x, we produce a 10000-valued vector L, with exactly 100 1s and the rest zeroes. 
+        This 'leaf indicator' vector can be considered the ensemble-induced 'footprint' of the example.</para>
+        <para>The 'distance' between two examples in the L-space is actually a Hamming distance, and is equal to the number of trees that do not distinguish the two examples.</para>
+        <para>We could repeat the same thought process for the non-leaf, or internal, nodes of the trees (we know that each tree has exactly 99 of them in our 100-leaf example), 
+        and produce another indicator vector, N (size 9900), for each example, indicating the 'trajectory' of each example through each of the trees.</para>
+        <para>The distance in the combined 19900-dimensional LN-space will be equal to the number of 'decisions' in all trees that 'agree' on the given pair of examples.</para>
+        <para>The TreeLeafFeaturizer is also producing the third vector, T, which is defined as Ti(x) = output of tree #i on example x.</para>
+      </remarks>
+      <example>
+        <code>
+          pipeline.Add(new TreeLeafFeaturizer())
```

Comment:
>pipeline.Add(new TreeLeafFeaturizer()) [](start = 10, length = 38)

This is very specific to the `CSharpApi.cs` code, and will be going away. We're going to need to remember to nuke it. Will this be going away?

Reply:
the comment would apply to all the examples. Keeping it, since leaving the documentation without any example would not look good. 

---
In reply to: [202459319](https://github.com/dotnet/machinelearning/pull/529#discussion_r202459319) [](ancestors = 202459319)